examples/tutorials/core_api_mnist/model_def_adaptive.py

```diff
@@ -0,0 +1,188 @@
+
+            if (batch_idx + 1) % args.log_interval == 0:
+                core_context.train.report_validation_metrics(steps_completed= (epoch) * len_trainset, \
+                 metrics={"test_loss": (test_loss /len(test_loader.dataset))})
+              
+    test_loss /= len(test_loader.dataset)
+
+    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
+        test_loss, correct, len(test_loader.dataset),
+        100. * correct / len(test_loader.dataset)))
+
+
+def load_state(checkpoint_directory):
+    checkpoint_directory = pathlib.Path(checkpoint_directory)
+    with checkpoint_directory.joinpath('checkpoint.pt').open("rb") as f:
+        return torch.load(f)
+
+
+def main(core_context):
+    # Training settings
+    parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
+    parser.add_argument('--batch-size', type=int, default=64, metavar='N',
+                        help='input batch size for training (default: 64)')
+    parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
+                        help='input batch size for testing (default: 1000)')
+    parser.add_argument('--epochs', type=int, default=10, metavar='N',
+                        help='number of epochs to train (default: 14)')
+    parser.add_argument('--lr', type=float, default=1.0, metavar='LR',
+                        help='learning rate (default: 1.0)')
+    parser.add_argument('--gamma', type=float, default=0.7, metavar='M',
+                        help='Learning rate step gamma (default: 0.7)')
+    parser.add_argument('--no-cuda', action='store_true', default=False,
+                        help='disables CUDA training')
+    parser.add_argument('--no-mps', action='store_true', default=True,
+                        help='disables macOS GPU training')
+    parser.add_argument('--dry-run', action='store_true', default=False,
+                        help='quickly check a single pass')                    
+    parser.add_argument('--seed', type=int, default=1, metavar='S',
+                        help='random seed (default: 1)')
+    parser.add_argument('--log-interval', type=int, default=10, metavar='N',
+                        help='how many batches to wait before logging training status')
+    parser.add_argument('--save-model', action='store_true', default=True,
+                        help='For Saving the current Model')
+  
+    args = parser.parse_args()
+    use_cuda = not args.no_cuda and torch.cuda.is_available()
+    use_mps = not args.no_mps and torch.backends.mps.is_available()
+
+    info = det.get_cluster_info()
+    assert info is not None, "this example only runs on-cluster"
+    latest_checkpoint = info.latest_checkpoint
+    if latest_checkpoint is not None: 
+        with core_context.checkpoint.restore_path(latest_checkpoint) as path:
+            model = load_state(path)
+
+    # NEW: get hparams from config
+    hparams = info.trial.hparams
```

Comment:
it looks like you create `hparams` here but don't use it until you create the `Net`?  I'd suggest setting the `hparams` variable down where you use it.

Also, don't say "get hparams from config", because technically you are not getting hparams from the experiment config.  The experiment config contains a hyperparameter _space_, that is, a bunch of ranges that these hyperparameters might take on.

But the `info.trial.hparams` is a _hyperparameter sample_, or one _point_ in that hyperparameter _space_, where each hparam has an exact value.  Does that distinction make sense?

So you're not "getting hparams from the config", you're "getting hparams chosen for this trial, from the cluster info object" or something like that.

Reply:
Yep, makes sense - thank you again for the explanation!